File: python/stack/valid_parentheses_neetcode_150.py
my_str = input()
opening = []
is_valid = True
for ch in my_str:
    if ch in "({[":
        opening.append(ch)
    elif ch in ")}]":
        if not opening:
            is_valid = False
            break
        top = opening.pop()
        if (ch == ')' and top != '(') or (ch == '}' and top != '{') or (ch == '[' and top != ']'):
            is_valid = False
        
    else:
        is_valid = False

print("The string is", is_valid and not opening)

File: rust/stack/valid_parentheses_neetcode_150.rs
#![allow(dead_code)]
use std::io;

#[derive(Debug)]
struct Stack<T> {
    stack: Vec<T>,
}

impl<T> Stack<T> {
    fn new() -> Self {
        Stack { stack: Vec::new() }
    }

    fn is_empty(&self) -> bool {
        self.stack.is_empty()
    }

    fn push(&mut self, item: T) {
        self.stack.push(item)
    }

    fn pop(&mut self) -> Option<T> {
        self.stack.pop()
    }

    fn top(&self) -> Option<&T> {
        self.stack.last()
    }
}

fn main() {
    let mut opening: Stack<char> = Stack::new();
    let mut is_valid = true;
    let mut input = String::new();
    io::stdin().read_line(&mut input).expect("oops");
    for ch in input.trim().chars() {
        if ch == '(' || ch == '{' || ch == '[' {
            opening.push(ch);
        } else if ch == ')' || ch == '}' || ch == ']' {
            if opening.is_empty() {
                is_valid = false;
                break;
            }
            let top = opening.pop().unwrap();
            if (ch == ')' && top != '(') || (ch == '}' && top != '{') || (ch == ']' && top != '[') {
                is_valid = false;
            }
        } else {
            is_valid = false;
        }
    }

    let answer = is_valid && opening.is_empty();
    println!("The string is valid: {}", answer);
}
